### ISSUE TITLE
Core dump when clearing insanely huge map

### ISSUE THREAD

**Issue by mattiwei** [haproxy/haproxy#419]
## Output of `haproxy -vv` and `uname -a`
```
./haproxy -vv
HA-Proxy version 2.0.11 2019/12/11 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits
  OPTIONS = USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_ZLIB=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT -PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL -LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
Running on OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.7
Running on zlib version : 1.2.7
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.23 2017-02-14
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE     mux=H2
              h2 : mode=HTTP       side=FE        mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services : none

Available filters :
        [SPOE] spoe
        [COMP] compression
        [CACHE] cache
        [TRACE] trace

 uname -a
Linux [REDACTED] 3.10.0-957.27.2.el7.x86_64 #1 SMP Tue Jul 9 16:53:14 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

```
## What's the configuration?

 ```
global
    stats socket ./stats.socket

frontend fe
    bind :6789
    mode tcp
    log-format [%t]\ %ci:%cp(%[var(sess.country)])->%[dst]:%[dst_port]
    tcp-request session set-var(sess.country) src,map_ip(./maps/10-8.map,--)
    default_backend be

backend be
    server localhost-8080 127.0.0.1:8080
```

## Steps to reproduce the behavior

1.  create a massive static map file
```
#!/usr/bin/perl

use strict;
use warnings;

for (my $b=0; $b <= 255; $b++) {
        for (my $c=0; $c <= 255; $c++) {
                for (my $d=0; $d <= 255; $d++) {
                        print "10.$b.$c.$d/32 10:$b:$c:$d\n";
                }
        }
}
```
execute above script using "perl ./generate-map.pl >10-8.map". This will create the file 10-8.map with 2^24 lines (16 mio lines, 510 MB file size). Copy the file to the destination directory specified in the configuration.

2. Fire up haproxy
```
./haproxy -f haproxy.cfg
```
Loading the map takes some seconds. Check with "top" when haproxy's CPU drops down again before doing the next step.

3. Delete the map using the stats socket from another terminal
```
echo "clear map ./maps/10-8.map" | socat ./stats.socket stdio
```

## Actual behavior
```
user@server01:~/haproxy/bin> ./haproxy -f haproxy.cfg
Thread 1 is about to kill the process.
*>Thread 1 : act=1 glob=0 wq=1 rq=1 tl=0 tlsz=0 rqsz=1
             stuck=1 fdcache=1 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=10673284473 now=12673840403 diff=2000555930
             curr_task=0x1113800b0 (task) calls=1 last=0
               fct=0x522960 (task_run_applet) ctx=0x11137ff00(<CLI>)
             strm=0x11137f960 src=unix fe=GLOBAL be=GLOBAL dst=<CLI>
             rqf=80c023 rqa=0 rpf=80008000 rpa=0 sif=EST,280008 sib=EST,204018
             af=(nil),0 csf=0x11137f910,1000
             ab=0x11137ff00,4 csb=(nil),0
             cof=0x11137f610,241300:PASS(0x11137f890)/RAW((nil))/unix_stream(19)
             cob=(nil),0:NONE((nil))/NONE((nil))/NONE(0)
  Thread 2 : act=0 glob=0 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
             stuck=0 fdcache=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=612159 now=660400 diff=48241
             curr_task=0
  Thread 3 : act=0 glob=0 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
             stuck=0 fdcache=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=699181 now=719842 diff=20661
             curr_task=0
  Thread 4 : act=0 glob=0 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
             stuck=0 fdcache=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=240574 now=291259 diff=50685
             curr_task=0
Aborted (core dumped)
```

## Expected behavior

haproxy should not dump a core.

## Do you have any idea what may have caused this?
(this is now a guess - I am not fluent in C...)
pat_ref_prune in pattern.c iterates through the whole elastic tree, deleting each entry one by one. My guess is that this takes longer as 2 seconds, and triggers a timeout

## Do you have an idea how to solve the issue?
(have mercy, these are my naive ideas)
- delete the whole tree in an atomic operation, not node by node
- impose a timeout (say 100ms) on pat_ref_prune. If the deletion has not finished yet when the timeout happens, call it again

**Comment by wtarreau** [comment]
Hi Mattiwei,

thanks a lot for your report. You're totally right. I don't know how
feasible it is to make the prune function yield, maybe it can have some
impacts if deleting some entries while traffic passes, but it certainly
is something we need to have a look at, as it's not acceptable to freeze
the whole process for two seconds! And by the way regarding consistency
during the reload operation, the new entries you're adding already force
yielding at least in order to read them, so it's very likely that atomicity
doesn't exist in that area anyway.

Willy

**Comment by wtarreau** [comment]
I'm starting to look at this. BTW, thanks for the reproducer, it definitely saves me time ;-)

**Comment by wtarreau** [comment]
@mattiwei I could come up with this. It's not perfect but at least it solves the issue for your case and is a way better improvement for me. I'm not backporting it yet, I'd like @thierryf78 to have a look at it and to check if there's any risk of unidentified corner case by yielding in the middle of the free operations. The code looks like it's already done in a sensible way so that any references are dropped before their contents are released so I guess it's OK, but we're never sure. I'm attaching the patch and am welcoming your feedback on it.
[0001-MEDIUM-map-make-the-clear-map-operation-yield.patch.txt](https://github.com/haproxy/haproxy/files/3989461/0001-MEDIUM-map-make-the-clear-map-operation-yield.patch.txt)

**Comment by mattiwei** [comment]
Hi Willy,
I tried the patch on Centos 8.0.1905 and RHEL 7.6 - I was not able to trigger a core dump anymore. 

Thank you very much for the super fast response! 

I'm taking the chance to extend my gratitude towards everybody involved in developing/documenting/testing/using haproxy - in short: being a part of the haproxy community (this includes everybody @haproxy.com). Joyeux noël /Frohe Weihnachten / Merry Christmas!
Mathias

@wtarreau : I don't close the issue yet as you wanted to check back with @thierryf78 . Feel free to close it whenever you want.

**Comment by wtarreau** [comment]
Excellent, thanks for testing. I suggest you to keep it in your patch set for now (at least it will save you from this trouble) and to report any issue you'd possibly observe that could be related (I'm aware of none obviously). This will definitely help in deciding whether to merge and backport it or not (or even to refine it). And yes, I'd definitely prefer to have Thierry's opinion before merging it.

**Comment by wtarreau** [comment]
Closing now since merged as d1d005d.
